contracts/external/nexus/WrappedCoveredPrincipalToken.sol

```diff
@@ -0,0 +1,145 @@
+    // a). Tranche should have same underlying/base token (i.e ETH, BTC, USDC).
+    // b). Should have the similar risk profiles.
+    EnumerableSet.AddressSet private _allowedTranches;
+
+    // Emitted when new tranche get whitelisted.
+    event TrancheAdded(address _tranche);
+
+    /// @notice Modifier to validate the tranche is whitelisted or not.
+    modifier isValidTranche(address _tranche) {
+        require(isAllowedTranche(_tranche), "WFP:ALREADY_EXISTS");
+        _;
+    }
+
+    ///@notice Initialize the wrapped token.
+    ///@dev    Wrapped token have 18 decimals, It is independent of the baseToken decimals.
+    constructor(address baseToken_, address owner_)
+        ERC20Permit(
+            _processName(IERC20Metadata(baseToken_).symbol()),
+            _processSymbol(IERC20Metadata(baseToken_).symbol())
+        )
+    {
+        baseToken = baseToken_;
+        _authorize(owner_);
+        setOwner(owner_);
+    }
+
+    ///@notice Allows to create the name for the wrapped token.
+    function _processName(string memory _tokenSymbol)
+        internal
+        pure
+        returns (string memory)
+    {
+        return
+            string(
+                abi.encodePacked("Wrapped", _tokenSymbol, "Futures Principal")
+            );
+    }
+
+    ///@notice Allows to create the symbol for the wrapped token.
+    function _processSymbol(string memory _tokenSymbol)
+        internal
+        pure
+        returns (string memory)
+    {
+        return string(abi.encodePacked("ep:", "W", _tokenSymbol));
+    }
+
+    /// @notice Add tranches within the allowed tranches enumerable set.
+    /// @dev    It is only allowed to execute by the owner of the contract.
+    ///         Tranches which has underlying token equals to the base token are
+    ///         only allowed to add, Otherwise it will revert.
+    /// @param  _tranche Address of the tranche which needs to add.
+    function addTranche(address _tranche)
+        external
+        isValidTranche(_tranche)
+        onlyOwner
+    {
+        require(
+            address(ITranche(_tranche).underlying()) == baseToken,
+            "WFP:INVALID_TRANCHE"
+        );
+        _allowedTranches.add(_tranche);
+        emit TrancheAdded(_tranche);
+    }
+
+    /// @notice Allows the defaulter to mint wrapped tokens (Covered position) by
+    ///         sending the de-pegged token to the contract.
+    /// @dev    a) Only allow minting the covered position when the tranche got expired otherwise revert.
+    ///         b) Sufficient allowance of the principal token (i.e tranche) should be provided
+    ///            to the contract by the `msg.sender` to make execution successful.
+    /// @param  _amount Amount of covered position / wrapped token `msg.sender` wants to mint.
+    /// @param  _tranche Address of the tranche which is covered by this covered position contract / wrapped token.
+    function mint(uint256 _amount, address _tranche)
```

Comment:
our principal tokens implement permit, might be worth adding permit functionality here to avoid a TX